src/typechecker/FStar.TypeChecker.Rel.fs

```diff
@@ -2601,20 +2635,28 @@ and solve_t_flex_flex env orig wl (lhs:flex_t) (rhs:flex_t) : solution =
                  in
                  let gamma_w = gamma_until u_lhs.ctx_uvar_gamma ctx_w in
                  let zs = intersect_binders gamma_w (ctx_l @ binders_lhs) (ctx_r @ binders_rhs) in
-                 let _, w, wl = new_uvar ("flex-flex quasi:"
+                 let new_uvar_typ = U.arrow zs (S.mk_Total t_res_lhs) in
+                 if snd (occurs u_lhs new_uvar_typ)
+                 ||  (not (Unionfind.equiv u_lhs.ctx_uvar_head u_rhs.ctx_uvar_head) &&
+                     snd (occurs u_rhs new_uvar_typ))
```

Comment:
Here's the main new check: If the type of the newly introduce variable mentions either one of the lhs/rhs uvars, then defer this constraint or fail.

Reply:
The rest of the code changes are really in support of checking if we are able to defer here, in case defer_ok=DeferAny or DeferFlexFlexOnly